sharding/contracts/sharding_manager.sol

```diff
@@ -162,31 +160,50 @@ contract SMC {
         emit NotaryReleased(notaryAddress, index);
     }
 
-    /// Calcuates the hash of the header from the input parameters
-    function computeHeaderHash(
-        uint256 shardId,
-        bytes32 parentHash,
-        bytes32 chunkRoot,
-        uint256 period,
-        address proposerAddress
-        ) public returns(bytes32) {
-      /*
-        TODO: Calculate the hash of the collation header from the input parameters
-      */
-    }
-
     /// Add collation header to the main chain, anyone can call this function. It emits a log
     function addHeader(
         uint _shardId,
-        uint period,
-        bytes32 chunkRoot,
-        address proposerAddress
+        uint _period,
+        bytes32 _chunkRoot
         ) public {
-      /*
-        TODO: Anyone can call this at any time. The first header
-        to get included for a given shard in a given period gets in,
-        all others don’t. This function just emits a log
-      */
+        require((_shardId >= 0) && (_shardId < SHARD_COUNT));
+        require(_period == block.number / PERIOD_LENGTH);
+        require(_period > lastSubmittedCollation[_shardId]);
+
+        //Track the numbers of participating notaries in between periods
+        //updateNotarySampleSize();
+
+        collationRecords[_shardId][_period] = CollationRecord({
+            chunkRoot: _chunkRoot,
+            proposer: msg.sender,
+            isElected: false
+        });
+
+        lastSubmittedCollation[_shardId] = block.number / PERIOD_LENGTH;
+        delete currentVote[_shardId];
+
+        emit HeaderAdded(_shardId, _chunkRoot, _period, msg.sender);
+    }
+
+    /// Sampled notary can call the following funtion to submit vote,
+    /// a vote log will be emitted for client to monitor
+    function submitVote(
+        uint _shardId,
+        uint _period,
+        uint _index,
+        bytes32 _chunkRoot        
+    ) public {
+        require(notaryRegistry[msg.sender].deposited);
+        require(collationRecords[_shardId][_period].chunkRoot == _chunkRoot);
+        require(!hasVoted(_shardId, _index));
+        require(getNotaryInCommittee(_shardId, _index) == msg.sender);
+
```

Comment:
It seems that we need to check `_period` against current period and `lastSubmittedCollation[_shardId]`.

Reply:
Also check against `_index` against `COMMITTEE_SIZE`.

Reply:
Thanks. Checks added